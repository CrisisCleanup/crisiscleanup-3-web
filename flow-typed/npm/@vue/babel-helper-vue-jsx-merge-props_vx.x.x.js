// flow-typed signature: d7a23c53f4091ef2bf1df8afa5eb43c4
// flow-typed version: <<STUB>>/@vue/babel-helper-vue-jsx-merge-props_v^1.0.0/flow_v0.131.0

/**
 * This is an autogenerated libdef stub for:
 *
 *   '@vue/babel-helper-vue-jsx-merge-props'
 *
 * Fill this stub out by replacing all the `any` types.
 *
 * Once filled out, we encourage you to share your work with the
 * community by sending a pull request to:
 * https://github.com/flowtype/flow-typed
 */

declare module '@vue/babel-helper-vue-jsx-merge-props' {
  declare module.exports: any;
}

/**
 * We include stubs for each file inside this npm package in case you need to
 * require those files directly. Feel free to delete any files that aren't
 * needed.
 */
declare module '@vue/babel-helper-vue-jsx-merge-props/dist/helper' {
  declare module.exports: any;
}

// Filename aliases
declare module '@vue/babel-helper-vue-jsx-merge-props/dist/helper.js' {
  declare module.exports: $Exports<'@vue/babel-helper-vue-jsx-merge-props/dist/helper'>;
}
