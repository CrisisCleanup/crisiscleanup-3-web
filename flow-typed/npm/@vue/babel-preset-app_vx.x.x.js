// flow-typed signature: 6ba91a92845d9a556a5bcdcc4f033fed
// flow-typed version: <<STUB>>/@vue/babel-preset-app_v^4.3.1/flow_v0.129.0

/**
 * This is an autogenerated libdef stub for:
 *
 *   '@vue/babel-preset-app'
 *
 * Fill this stub out by replacing all the `any` types.
 *
 * Once filled out, we encourage you to share your work with the
 * community by sending a pull request to:
 * https://github.com/flowtype/flow-typed
 */

declare module '@vue/babel-preset-app' {
  declare module.exports: any;
}

/**
 * We include stubs for each file inside this npm package in case you need to
 * require those files directly. Feel free to delete any files that aren't
 * needed.
 */
declare module '@vue/babel-preset-app/polyfillsPlugin' {
  declare module.exports: any;
}

// Filename aliases
declare module '@vue/babel-preset-app/index' {
  declare module.exports: $Exports<'@vue/babel-preset-app'>;
}
declare module '@vue/babel-preset-app/index.js' {
  declare module.exports: $Exports<'@vue/babel-preset-app'>;
}
declare module '@vue/babel-preset-app/polyfillsPlugin.js' {
  declare module.exports: $Exports<'@vue/babel-preset-app/polyfillsPlugin'>;
}
