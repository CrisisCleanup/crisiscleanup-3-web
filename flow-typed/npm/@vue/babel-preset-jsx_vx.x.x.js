// flow-typed signature: eeecaeff29e4a3a6c63c4fe7e16a0eb3
// flow-typed version: <<STUB>>/@vue/babel-preset-jsx_v^1.1.2/flow_v0.131.0

/**
 * This is an autogenerated libdef stub for:
 *
 *   '@vue/babel-preset-jsx'
 *
 * Fill this stub out by replacing all the `any` types.
 *
 * Once filled out, we encourage you to share your work with the
 * community by sending a pull request to:
 * https://github.com/flowtype/flow-typed
 */

declare module '@vue/babel-preset-jsx' {
  declare module.exports: any;
}

/**
 * We include stubs for each file inside this npm package in case you need to
 * require those files directly. Feel free to delete any files that aren't
 * needed.
 */
declare module '@vue/babel-preset-jsx/dist/plugin.cjs' {
  declare module.exports: any;
}

// Filename aliases
declare module '@vue/babel-preset-jsx/dist/plugin.cjs.js' {
  declare module.exports: $Exports<'@vue/babel-preset-jsx/dist/plugin.cjs'>;
}
