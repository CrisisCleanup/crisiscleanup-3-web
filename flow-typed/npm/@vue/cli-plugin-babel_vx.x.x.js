// flow-typed signature: 4cf1a98e80b05c0a167e013d64bc9645
// flow-typed version: <<STUB>>/@vue/cli-plugin-babel_v^4.4.6/flow_v0.131.0

/**
 * This is an autogenerated libdef stub for:
 *
 *   '@vue/cli-plugin-babel'
 *
 * Fill this stub out by replacing all the `any` types.
 *
 * Once filled out, we encourage you to share your work with the
 * community by sending a pull request to:
 * https://github.com/flowtype/flow-typed
 */

declare module '@vue/cli-plugin-babel' {
  declare module.exports: any;
}

/**
 * We include stubs for each file inside this npm package in case you need to
 * require those files directly. Feel free to delete any files that aren't
 * needed.
 */
declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/customConfig.input' {
  declare module.exports: any;
}

declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/customConfig.output' {
  declare module.exports: any;
}

declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/default.input' {
  declare module.exports: any;
}

declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/default.output' {
  declare module.exports: any;
}

declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/doubleQuote.input' {
  declare module.exports: any;
}

declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/doubleQuote.output' {
  declare module.exports: any;
}

declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/require.input' {
  declare module.exports: any;
}

declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/require.output' {
  declare module.exports: any;
}

declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/templateLiteral.input' {
  declare module.exports: any;
}

declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/templateLiteral.output' {
  declare module.exports: any;
}

declare module '@vue/cli-plugin-babel/codemods/usePluginPreset' {
  declare module.exports: any;
}

declare module '@vue/cli-plugin-babel/generator' {
  declare module.exports: any;
}

declare module '@vue/cli-plugin-babel/migrator' {
  declare module.exports: any;
}

declare module '@vue/cli-plugin-babel/preset' {
  declare module.exports: any;
}

// Filename aliases
declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/customConfig.input.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel/codemods/__testfixtures__/customConfig.input'>;
}
declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/customConfig.output.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel/codemods/__testfixtures__/customConfig.output'>;
}
declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/default.input.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel/codemods/__testfixtures__/default.input'>;
}
declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/default.output.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel/codemods/__testfixtures__/default.output'>;
}
declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/doubleQuote.input.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel/codemods/__testfixtures__/doubleQuote.input'>;
}
declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/doubleQuote.output.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel/codemods/__testfixtures__/doubleQuote.output'>;
}
declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/require.input.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel/codemods/__testfixtures__/require.input'>;
}
declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/require.output.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel/codemods/__testfixtures__/require.output'>;
}
declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/templateLiteral.input.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel/codemods/__testfixtures__/templateLiteral.input'>;
}
declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/templateLiteral.output.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel/codemods/__testfixtures__/templateLiteral.output'>;
}
declare module '@vue/cli-plugin-babel/codemods/usePluginPreset.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel/codemods/usePluginPreset'>;
}
declare module '@vue/cli-plugin-babel/generator.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel/generator'>;
}
declare module '@vue/cli-plugin-babel/index' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel'>;
}
declare module '@vue/cli-plugin-babel/index.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel'>;
}
declare module '@vue/cli-plugin-babel/migrator/index' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel/migrator'>;
}
declare module '@vue/cli-plugin-babel/migrator/index.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel/migrator'>;
}
declare module '@vue/cli-plugin-babel/preset.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel/preset'>;
}
