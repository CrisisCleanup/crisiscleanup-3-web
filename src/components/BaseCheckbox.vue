<template>
  <label class="checkbox-container">
    <slot></slot>
    <input
      type="checkbox"
      :checked="value"
      class="checkmark-input"
      @input="$emit('input', $event.target.checked)"
      @change="$emit('change', $event.target.checked)"
    />
    <span class="checkmark"></span>
  </label>
</template>
<script>
export default {
  name: 'BaseCheckbox',
  props: ['value'],
};
</script>

<style scoped>
/* Customize the label (the container) */
.checkbox-container {
  display: block;
  position: relative;
  padding-left: 30px;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

/* Hide the browser's default checkbox */
.checkbox-container input {
  position: absolute;
  opacity: 0;
  cursor: pointer;
  height: 0;
  width: 0;
}

/* Create a custom checkbox */
.checkmark {
  position: absolute;
  top: 0;
  left: 0;
  height: 20px;
  width: 20px;
  border: solid 1px #c8cacf;
  background-color: white;
}

/* When the checkbox is checked, add a blue background */
.checkbox-container input:checked ~ .checkmark {
  background-color: #000000;
}

.checkbox-container input:disabled ~ .checkmark {
  background-color: #e4e5e7;
}

/* Create the checkmark/indicator (hidden when not checked) */
.checkmark:after {
  content: '';
  position: absolute;
  display: none;
}

/* Show the checkmark when checked */
.checkbox-container input:checked ~ .checkmark:after {
  display: block;
}

/* Style the checkmark/indicator */
.checkbox-container .checkmark:after {
  left: 7px;
  top: 3px;
  width: 5px;
  height: 10px;
  border: solid white;
  border-width: 0 2px 2px 0;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
}
</style>
