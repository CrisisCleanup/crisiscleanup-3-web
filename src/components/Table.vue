<template>
  <div class="table-grid w-full">
    <div
      v-if="!hideHeader && $mq !== 'sm'"
      class="header text-crisiscleanup-grey-700 bg-white"
      :style="gridStyleHeader"
    >
      <div
        v-if="enableSelection && $mq !== 'sm'"
        class="flex items-center p-2 border-b"
      >
        <base-checkbox class="mb-5" @input="setAllChecked" />
      </div>
      <div
        v-if="hasRowDetails && $mq !== 'sm'"
        class="flex items-center p-2 border-b"
      ></div>
      <div
        v-for="column of columns"
        :key="column.key"
        class="p-2 border-b flex items-center cursor-pointer header-column"
        :class="column.headerClass || []"
        @click="
          () => {
            if (column.sortable) {
              sort(column.sortKey || column.key);
            }
          }
        "
      >
        <slot :name="`${column.key}-title`" :column="column">
          <base-text
            class="text-crisiscleanup-grey-700"
            :class="column.titleClass && column.titleClass"
            variant="h3"
            regular
          >
            {{ $t(column.title) }}
          </base-text>
        </slot>
        <div v-if="column.sortable">
          <ccu-icon
            v-if="
              sorter.key === (column.sortKey || column.key) &&
              sorter.direction === 'asc'
            "
            :alt="$t('actions.sort_ascending')"
            size="small"
            type="up"
          />
          <ccu-icon
            v-else-if="
              sorter.key === (column.sortKey || column.key) &&
              sorter.direction === 'desc'
            "
            :alt="$t('actions.sort_descending')"
            size="small"
            type="down"
          />
          <ccu-icon
            v-else
            :alt="$t('actions.sortable')"
            size="small"
            type="updown"
          />
        </div>
      </div>
      <template v-if="enableColumnSearch">
        <div
          v-for="column of columns"
          :key="`search-${column.key}`"
          class="
            p-2
            border-b
            flex
            items-center
            cursor-pointer
            bg-crisiscleanup-light-grey
          "
        >
          <template v-if="column.searchable">
            <form-select
              v-if="column.searchSelect"
              class="w-64 bg-white border h-10 border-crisiscleanup-dark-100"
              :options="column ? column.getSelectValues(data) : []"
              item-key="value"
              label="name_t"
              :value="columnSearch[column.key]"
              @input="
                (value) => {
                  columnSearch[column.key] = value;
                  onSearch();
                }
              "
            ></form-select>
            <base-input
              v-else
              :placeholder="column.title"
              :value="columnSearch[column.key]"
              @input="
                (value) => {
                  columnSearch[column.key] = value;
                  onSearch();
                }
              "
              input-style="width: 100%"
            ></base-input>
          </template>
        </div>
      </template>
    </div>
    <div class="body bg-white relative" :style="gridStyleBody">
      <div
        v-if="loading"
        class="
          absolute
          bottom-0
          left-0
          right-0
          top-0
          bg-crisiscleanup-light-grey
          opacity-75
          flex
          items-center
          justify-center
        "
      >
        <spinner />
      </div>
      <div
        v-for="item of data"
        :key="item.id"
        :style="gridStyleRow"
        class="hover:bg-crisiscleanup-light-grey"
        :class="{ 'bg-crisiscleanup-light-grey': selectedItems.has(item.id) }"
        @click="rowClick(item, $event)"
      >
        <div
          v-if="enableSelection && $mq !== 'sm'"
          class="flex items-center p-2 lg:border-b md:border-b"
        >
          <base-checkbox
            :value="selectedItems.has(item.id)"
            class="mb-5"
            data-cy="tableview_actionSelect"
            @input="
              (value) => {
                setChecked(item, value);
              }
            "
          />
        </div>
        <div
          v-if="hasRowDetails && $mq !== 'sm'"
          class="flex items-center p-2 lg:border-b md:border-b"
        >
          <font-awesome-icon
            class="cursor-pointer"
            size="md"
            :icon="showingDetails.has(item.id) ? 'caret-up' : 'caret-down'"
            @click="setShowingDetails(item)"
          />
        </div>
        <div
          v-for="column of columns"
          :key="column.key"
          class="flex items-center p-2 lg:border-b md:border-b cursor-pointer"
          :class="column.class || []"
          @click="handleColumnAction(column, item[column.key], item)"
        >
          <slot :name="column.key" :item="item">
            <span v-if="$mq === 'sm'" class="font-semibold mr-2">
              {{ column.title }}:
            </span>
            <template
              v-if="
                item[column.key] || item[column.key] === 0 || column.transformer
              "
            >
              <span v-if="column.transformer">{{
                column.transformer(item[column.key], item)
              }}</span>
              <span v-else>{{
                column.subKey
                  ? item[column.key][column.subKey]
                  : item[column.key]
              }}</span>
            </template>
          </slot>
        </div>
        <div
          v-if="hasRowDetails"
          v-show="showingDetails.has(item.id)"
          :style="`grid-column-start: 1; grid-column-end: ${
            columns.length + 1
          };`"
        >
          <slot name="rowDetails" :item="item"></slot>
        </div>
      </div>
      <div v-if="!data.length" class="p-2 text-crisiscleanup-grey-700 italic">
        {{ $t('info.no_items_found') }}
      </div>
    </div>
    <div
      v-if="enablePagination"
      class="footer flex items-center justify-between p-4"
    >
      <div class="flex items-center">
        <span class="mr-2">{{ $t('tableVue.per_page') }}</span>
        <form-select
          :value="pagination.pageSize"
          :options="pageSizes"
          :clearable="false"
          select-classes="w-24 bg-white border vue-select-up"
          @input="onSelectPageSize"
        />
      </div>
      <div>
        <div class="flex items-center">
          <base-button
            :disabled="isPreviousButtonDisabled"
            :action="
              () => {
                pageChangeHandle('previous');
              }
            "
            :text="$t('tableVue.prev')"
            :alt="$t('tableVue.prev')"
            icon="caret-left"
            class="mr-3 text-base"
          />
          <div>
            <template v-for="trigger in paginationTriggers">
              <span
                :key="trigger"
                :class="{
                  'rounded-full border px-3 py-1 bg-white shadow-inner':
                    trigger === pagination.current,
                }"
                class="cursor-pointer mx-4 text-base"
                @click="
                  () => {
                    pageChangeHandle(trigger);
                  }
                "
                >{{ trigger }}</span
              >
            </template>
          </div>
          <base-button
            :disabled="isNextButtonDisabled"
            :action="
              () => {
                pageChangeHandle('next');
              }
            "
            :text="$t('tableVue.next')"
            :alt="$t('tableVue.next')"
            suffix-icon="caret-right"
            class="text-base"
          />
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import { EventsMixin } from '@/mixins';
import { exportCSVFile } from '../utils/downloads';

export default {
  name: 'Table',
  mixins: [EventsMixin],
  props: {
    columns: {
      type: Array,
      default: () => {
        return [];
      },
    },
    hideHeader: {
      type: Boolean,
    },
    data: {
      type: null,
      default: null,
    },
    pagination: {
      type: Object,
      default: () => {
        return {};
      },
    },
    sorter: {
      type: Object,
      default: () => {
        return {};
      },
    },
    columnSearch: {
      type: Object,
      default: () => {
        return {};
      },
    },
    loading: Boolean,
    enableSelection: Boolean,
    enablePagination: Boolean,
    enableColumnSearch: Boolean,
    hasRowDetails: Boolean,
    bodyStyle: {
      type: Object,
      default: () => {
        return {};
      },
    },
  },
  data() {
    return {
      pageSizes: [5, 10, 20, 100, 500, 1000],
      visiblePagesCount: 5,
      selectedItems: new Set([]),
      showingDetails: new Set([]),
    };
  },
  computed: {
    pageCount() {
      return Math.ceil((this.pagination.total || 1) / this.pagination.pageSize);
    },
    isPreviousButtonDisabled() {
      return this.pagination.current === 1;
    },
    isNextButtonDisabled() {
      return this.pagination.current === this.pageCount;
    },
    paginationTriggers() {
      const currentPage = this.pagination.current;
      const { pageCount } = this;
      const { visiblePagesCount } = this;
      const visiblePagesThreshold = (visiblePagesCount - 1) / 2;
      const pagintationTriggersArray = Array(this.visiblePagesCount - 1).fill(
        0,
      );

      if (pageCount < visiblePagesCount) {
        return Array(pageCount)
          .fill(0)
          .map((_, index) => index + 1);
      }

      if (currentPage <= visiblePagesThreshold + 1) {
        pagintationTriggersArray[0] = 1;
        const pagintationTriggers = pagintationTriggersArray.map(
          (paginationTrigger, index) => {
            return pagintationTriggersArray[0] + index;
          },
        );
        pagintationTriggers.push(pageCount);
        return pagintationTriggers;
      }

      if (currentPage >= pageCount - visiblePagesThreshold + 1) {
        const pagintationTriggers = pagintationTriggersArray.map(
          (paginationTrigger, index) => {
            return pageCount - index;
          },
        );
        pagintationTriggers.reverse().unshift(1);
        return pagintationTriggers;
      }

      pagintationTriggersArray[0] = currentPage - visiblePagesThreshold + 1;
      const pagintationTriggers = pagintationTriggersArray.map(
        (paginationTrigger, index) => {
          return pagintationTriggersArray[0] + index;
        },
      );
      pagintationTriggers.unshift(1);
      pagintationTriggers[pagintationTriggers.length - 1] = pageCount;
      return pagintationTriggers;
    },
    gridTemplate() {
      const columnWidths = this.columns.map((column) => column.width || '1fr');
      if (this.hasRowDetails) {
        columnWidths.unshift('35px');
      }
      if (this.enableSelection) {
        columnWidths.unshift('35px');
      }
      return columnWidths.join(' ');
    },
    gridStyleHeader() {
      return {
        display: 'grid',
        'grid-template-columns': this.gridTemplate,
      };
    },
    gridStyleBody() {
      return {
        'max-height': '600px',
        overflow: 'auto',
        ...this.bodyStyle,
      };
    },
    gridStyleRow() {
      return {
        display: 'grid',
        'grid-template-columns': this.$mq === 'sm' ? 'auto' : this.gridTemplate,
      };
    },
  },
  mounted() {},
  methods: {
    setChecked(item, value) {
      if (value) {
        this.selectedItems.add(item.id);
      } else {
        this.selectedItems.delete(item.id);
      }
      this.selectedItems = new Set(this.selectedItems);
      this.$emit('selectionChanged', this.selectedItems);
    },
    setShowingDetails(item) {
      if (this.showingDetails.has(item.id)) {
        this.showingDetails.delete(item.id);
      } else {
        this.showingDetails.add(item.id);
      }
      this.showingDetails = new Set(this.showingDetails);
    },
    setAllChecked(value) {
      if (value) {
        this.selectedItems = new Set(this.data.map((item) => item.id));
        this.$emit('allSelected', this.selectedItems);
      } else {
        this.selectedItems = new Set();
        this.$emit('allDeselected', this.selectedItems);
      }
      this.$emit('selectionChanged', this.selectedItems);
    },
    pageChangeHandle(value) {
      let newPage;
      switch (value) {
        case 'next':
          newPage = this.pagination.current + 1;
          break;
        case 'previous':
          newPage = this.pagination.current - 1;
          break;
        default:
          newPage = value;
      }
      const columnSearch = { ...this.columnSearch };

      const pagination = {
        ...this.pagination,
        current: newPage,
        page: newPage,
      };
      const filter = {};
      const sorter = {
        ...this.sorter,
      };
      this.$emit('change', { pagination, filter, sorter, columnSearch });
    },
    onSelectPageSize(pageSize) {
      const columnSearch = { ...this.columnSearch };
      const pagination = {
        ...this.pagination,
        current: 1,
        pageSize,
      };
      const filter = {};
      const sorter = {
        ...this.sorter,
      };
      this.$emit('change', { pagination, filter, sorter, columnSearch });
    },
    sort(key) {
      const columnSearch = { ...this.columnSearch };
      const sorter = { ...this.sorter };
      sorter.key = key;
      if (sorter.direction === 'asc') {
        sorter.direction = 'desc';
      } else {
        sorter.direction = 'asc';
      }

      const pagination = {
        ...this.pagination,
      };
      const filter = {};
      this.logEvent(
        sorter.direction === 'asc' ? 'user_ui-sort-asc' : 'user_ui-sort-desc',
      );
      this.$emit('change', { pagination, filter, sorter, columnSearch });
    },
    onSearch() {
      const columnSearch = { ...this.columnSearch };
      const sorter = { ...this.sorter };
      const pagination = {
        ...this.pagination,
      };
      const filter = {};
      this.$emit('change', { pagination, filter, sorter, columnSearch });
    },
    rowClick(item, event) {
      const hasClass = (className) =>
        event.target.className.includes &&
        event.target.className.includes(className);
      if (
        event &&
        [
          'case-flag',
          'checkmark',
          'checkmark-input',
          'base-button',
          'link',
        ].some(hasClass)
      ) {
        return;
      }
      this.$emit('rowClick', item);
    },
    exportTableCSV() {
      const headers = {};
      const headersColumns = this.columns.filter((column) =>
        Boolean(column.title),
      );
      headersColumns.forEach((column) => {
        headers[column.key] = column.title;
      });

      const data = this.data.map((obj) => {
        const response = {};
        headersColumns.forEach((column) => {
          response[column.key] = obj[column.key];
        });
        return response;
      });
      exportCSVFile(headers, data, 'export');
    },
    async handleColumnAction(column, data, item) {
      if (column.action) {
        await column.action(data, item);
      }
    },
  },
};
</script>

<style scoped></style>
