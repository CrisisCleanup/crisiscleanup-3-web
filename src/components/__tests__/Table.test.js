/**
 *
 * Tests for Table
 *
 * Components
 */

import { shallowMount } from '@vue/test-utils';
import BaseCheckbox from '../BaseCheckbox';
import Table from '../Table';

const TestColumns = [
  {
    title: 'One',
    dataIndex: 'one_index',
    key: 'one_index',
    width: '0.5fr',
  },
  {
    title: 'Two',
    dataIndex: 'two_index',
    key: 'two_index',
    width: '1.5fr',
  },
];

const TestData = [{ id: 'one_index' }, { id: 'two_index' }];

const mountWithOptions = ({ props } = {}) =>
  shallowMount(Table, {
    stubs: {
      'base-checkbox': BaseCheckbox,
      'base-select': "<input type='select'>",
      'base-button': '<button />',
    },
    propsData: {
      columns: TestColumns,
      data: TestData,
      ...props,
    },
  });

describe('Table', () => {
  it('should not log any errors', () => {
    const spy = jest.spyOn(global.console, 'error');
    mountWithOptions();
    expect(spy).not.toHaveBeenCalled();
  });

  it('should update selected items', () => {
    const wrapper = mountWithOptions();
    wrapper.vm.setChecked({ id: 'one_index' }, true);
    expect(wrapper.vm.selectedItems).toMatchSnapshot();
    wrapper.vm.setChecked({ id: 'one_index' });
    wrapper.vm.setChecked({ id: 'one_index' }, false);
    expect(wrapper.vm.selectedItems).toMatchSnapshot();
  });

  it('should select all items', () => {
    const wrapper = mountWithOptions();
    wrapper.vm.setAllChecked(true);
    expect(wrapper.vm.selectedItems).toEqual(
      new Set(['one_index', 'two_index']),
    );
    wrapper.vm.setAllChecked();
    expect(wrapper.vm.selectedItems).toEqual(new Set([]));
  });

  it('should render correctly with selection and pagination', () => {
    const wrapper = mountWithOptions({
      props: {
        enablePagination: true,
        enableSelection: true,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should emit clicked item', () => {
    const wrapper = mountWithOptions();
    wrapper.vm.rowClick({ id: 1 });
    expect(wrapper.emitted().rowClick[0][0]).toEqual({ id: 1 });
  });

  it('should handle page change', () => {
    const wrapper = mountWithOptions({
      props: {
        // TODO: fix prop name typo
        enablePagniation: true,
        enableSelection: true,
        pagination: {
          pageSize: 5,
          page: 1,
          current: 1,
        },
      },
    });
    expect(wrapper.vm.pagination.current).toBe(1);
    wrapper.vm.pageChangeHandle('next');
    expect(wrapper.vm.pagination).toMatchSnapshot();
    wrapper.vm.pageChangeHandle('previous');
    expect(wrapper.vm.pagination).toMatchSnapshot();
    wrapper.vm.pageChangeHandle(10);
    expect(wrapper.vm.pagination).toMatchSnapshot();
  });

  it('should handle page size change', () => {
    const wrapper = mountWithOptions({
      props: {
        // TODO: fix prop name typo
        enablePagniation: true,
        enableSelection: true,
        pagination: {
          pageSize: 5,
          page: 1,
          current: 1,
        },
      },
    });
    wrapper.vm.onSelectPageSize(10);
    expect(wrapper.vm.pagination).toMatchSnapshot();
    wrapper.vm.onSelectPageSize(1000);
    expect(wrapper.vm.pagination).toMatchSnapshot();
  });

  it('should render correctly and match snapshot', () => {
    const wrapper = mountWithOptions();
    expect(wrapper.element).toMatchSnapshot();
  });
});
