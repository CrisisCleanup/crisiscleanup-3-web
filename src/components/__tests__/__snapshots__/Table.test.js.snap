// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should handle page change 1`] = `
Object {
  "current": 1,
  "page": 1,
  "pageSize": 5,
}
`;

exports[`Table should handle page change 2`] = `
Object {
  "current": 1,
  "page": 1,
  "pageSize": 5,
}
`;

exports[`Table should handle page change 3`] = `
Object {
  "current": 1,
  "page": 1,
  "pageSize": 5,
}
`;

exports[`Table should handle page size change 1`] = `
Object {
  "current": 1,
  "page": 1,
  "pageSize": 5,
}
`;

exports[`Table should handle page size change 2`] = `
Object {
  "current": 1,
  "page": 1,
  "pageSize": 5,
}
`;

exports[`Table should handle sort correctly from cold 1`] = `
Object {
  "current": 1,
  "page": 1,
  "pageSize": 5,
}
`;

exports[`Table should handle sort correctly with existing sort 1`] = `
Object {
  "current": 1,
  "page": 1,
  "pageSize": 5,
}
`;

exports[`Table should render correctly and match snapshot 1`] = `
<div
  class="table w-full"
>
  <div
    class="header text-crisiscleanup-grey-700 bg-white"
    style="display: grid; grid-template-columns: 0.5fr 1.5fr;"
  >
    <!---->

    <div
      class="p-2 border-b flex items-center cursor-pointer"
    >

      One

      <!---->
    </div>
    <div
      class="p-2 border-b flex items-center cursor-pointer"
    >

      Two

      <!---->
    </div>
  </div>

  <div
    class="body bg-white relative"
    style="max-height: 600px; overflow: auto;"
  >
    <!---->

    <div
      class="hover:bg-crisiscleanup-light-grey"
      style="display: grid; grid-template-columns: 0.5fr 1.5fr;"
    >
      <!---->

      <div
        class="flex items-center p-2 border-b cursor-pointer"
      >
        <!---->
      </div>
      <div
        class="flex items-center p-2 border-b cursor-pointer"
      >
        <!---->
      </div>
    </div>
    <div
      class="hover:bg-crisiscleanup-light-grey"
      style="display: grid; grid-template-columns: 0.5fr 1.5fr;"
    >
      <!---->

      <div
        class="flex items-center p-2 border-b cursor-pointer"
      >
        <!---->
      </div>
      <div
        class="flex items-center p-2 border-b cursor-pointer"
      >
        <!---->
      </div>
    </div>
  </div>

  <!---->
</div>
`;

exports[`Table should render correctly with selection and pagination 1`] = `
<div
  class="table w-full"
  enablepagination="true"
>
  <div
    class="header text-crisiscleanup-grey-700 bg-white"
    style="display: grid; grid-template-columns: 35px 0.5fr 1.5fr;"
  >
    <div
      class="flex items-center p-2 border-b"
    >
      <label
        class="checkbox-container mb-5"
      >

        <input
          class="checkmark-input"
          type="checkbox"
        />

        <span
          class="checkmark"
        />
      </label>
    </div>

    <div
      class="p-2 border-b flex items-center cursor-pointer"
    >

      One

      <!---->
    </div>
    <div
      class="p-2 border-b flex items-center cursor-pointer"
    >

      Two

      <!---->
    </div>
  </div>

  <div
    class="body bg-white relative"
    style="max-height: 600px; overflow: auto;"
  >
    <!---->

    <div
      class="hover:bg-crisiscleanup-light-grey"
      style="display: grid; grid-template-columns: 35px 0.5fr 1.5fr;"
    >
      <div
        class="flex items-center p-2 border-b"
      >
        <label
          class="checkbox-container mb-5"
          data-cy="tableview_actionSelect"
        >

          <input
            class="checkmark-input"
            type="checkbox"
          />

          <span
            class="checkmark"
          />
        </label>
      </div>

      <div
        class="flex items-center p-2 border-b cursor-pointer"
      >
        <!---->
      </div>
      <div
        class="flex items-center p-2 border-b cursor-pointer"
      >
        <!---->
      </div>
    </div>
    <div
      class="hover:bg-crisiscleanup-light-grey"
      style="display: grid; grid-template-columns: 35px 0.5fr 1.5fr;"
    >
      <div
        class="flex items-center p-2 border-b"
      >
        <label
          class="checkbox-container mb-5"
          data-cy="tableview_actionSelect"
        >

          <input
            class="checkmark-input"
            type="checkbox"
          />

          <span
            class="checkmark"
          />
        </label>
      </div>

      <div
        class="flex items-center p-2 border-b cursor-pointer"
      >
        <!---->
      </div>
      <div
        class="flex items-center p-2 border-b cursor-pointer"
      >
        <!---->
      </div>
    </div>
  </div>

  <!---->
</div>
`;

exports[`Table should update selected items 1`] = `
Set {
  "one_index",
}
`;

exports[`Table should update selected items 2`] = `Set {}`;
